Relax dimension ordering rules for dot_general.

hawkinsp opened the pull request:
JAX currently requires that batch dimensions appear first and contiguously in the arguments to dot_general. However, XLA does not require this; relax JAX's checks so that it also allows batch dimensions in arbitrary positions.

Since batch dimensions are now allowed in arbitrary positions, it's not hard to
generalize the dot_general batching rule to avoid performing any transposes. Fixes #2972 .

In passing, also move the bool/int dot expansion into the XLA translation rule. The expansion inside the `lax.dot_general()` wrapper predated the existence of (or at least my knowledge of) `xla.lower_fun()`.

hawkinsp committed:
Relax dimension ordering rules for dot_general.

JAX currently requires that batch dimensions appear first and contiguously in the arguments to dot_general. However, XLA does not require this; relax JAX's checks so that it also allows batch dimensions in arbitrary positions.

Since batch dimensions are now allowed in arbitrary positions, it's not hard to
generalize the dot_general batching rule to avoid performing any transposes
(#2972).

In passing, also move the bool/int dot expansion into the XLA translation rule. The expansion inside the `lax.dot_general()` wrapper predated the existence of (or at least my knowledge of) `xla.lower_fun()`.

jekbradbury reviewed (approved):
Thanks! I'm sorry I dropped the ball on this.

Also, last year I added a test [here](https://github.com/google/jax/pull/1331/files#diff-60f1c5fe94df9c63c0ea8aec5a29584e) to assert that vmap didn't add any broadcasts in certain cases; if vmap can now always take one `dot_general` to one `dot_general` we could add an assertion to that effect in the tests in `lax_vmap_test` instead (and delete the other one).

hawkinsp committed:
Also test for transpose in dot vmap test.

hawkinsp commented:
> Thanks! I'm sorry I dropped the ball on this.
> 
> Also, last year I added a test [here](https://github.com/google/jax/pull/1331/files#diff-60f1c5fe94df9c63c0ea8aec5a29584e) to assert that vmap didn't add any broadcasts in certain cases; if vmap can now always take one `dot_general` to one `dot_general` we could add an assertion to that effect in the tests in `lax_vmap_test` instead (and delete the other one).

The jaxpr has some call structure to it so instead I added an assertion that `transpose` doesn't appear in the jaxpr, rather than testing for a particular structure.

hawkinsp committed:
Use iteration over equations to test for "transpose" and "broadcast".
